### PR TITLE
Flash: Add serialization support

### LINKED ISSUE #1741: Serialize & load Flash metadata to disk
We need to store the channel metadata to disk and make sure to load it when we start up the flash nodes.

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
It's missing a test-case which I'll add soon.

Took @omerfirmak's advice and used occasional dumping rather than making it timer-based. I also used a template mixin trick so most of the existing doesn't need to be changed.

Fixes #1741

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Need to add `UpdateSigner` serialization too. Working on it.

**Comment by AndrejMitrovic** [comment]
I realize now that I could have used UDAs for this..

**Comment by AndrejMitrovic** [comment]
Fixed serialization issues and added a test-case. There were lots of little details that needed to be taken into account.

**Comment by AndrejMitrovic** [comment]
I'll file a separate issue for converting the mixin into just using UDAs for serialization. And also the need to add hashmap support to the serializer.

File: source/agora/flash/Node.d [hunk 1]
```diff
@@ -812,9 +941,9 @@ public abstract class FlashNode : ControlFlashAPI
         assert(result.error == ErrorCode.None, result.to!string);
 
         auto channel = new Channel(chan_conf, this.kp, priv_nonce, result.value,
-            peer, this.engine, this.taskman, &this.putTransaction,
```
Comment (omerfirmak) [inline source/agora/flash/Node.d:843]: I am not seeing why we need this extra `peer_pk` parameter. Is there something I am missing?

Comment (AndrejMitrovic) [reply source/agora/flash/Node.d:843]: Maybe not! I'll try to remove it. I think I needed it briefly during some refactorings.

**Comment by omerfirmak** [comment]
Could we have any reentrancy issues with `dump()`? I guess `ManagedDatabase` would handle any edge cases.

**Comment by AndrejMitrovic** [comment]
> Could we have any reentrancy issues with dump()? 

With the current assumption of only having one work thread the function will never be entered twice. The routine doesn't yield the fiber either. Do you see another way it might be entered twice?

**Comment by AndrejMitrovic** [comment]
Removed `peer_pk` from the ctor.

File: source/agora/flash/Channel.d [hunk 2]
```diff
@@ -1884,3 +1766,134 @@ LOuter: while (1)
         return this.pending_close.tx;
     }
 }
+
+/// The closing transaction can spend from the funding transaction when
+/// the channel parties want to collaboratively close the channel.
+/// It requires both of the parties signatures. For safety reasons,
+/// the closing transaction should only be signed once the node marks
+/// the channel as 'PendingClose'.
+private static struct PendingClose
```
Comment (Geod24) [inline source/agora/flash/Channel.d]: `static` no more needed

File: source/agora/flash/Channel.d [hunk 3]
```diff
@@ -1884,3 +1766,134 @@ LOuter: while (1)
...
+{
+    private Hash payment_hash;
+    private Amount amount;
+    private Height lock_height;
+    private OnionPacket packet;
+    private Height height;
+}
+
+/// All the channel metadata which we keep in the DB for storage
+private mixin template ChannelMetadata ()
```
Comment (Geod24) [inline source/agora/flash/Channel.d:2046]: This feels weird to me at first, that we have to use a mixin template here.

**Review by Geod24 — COMMENTED** [review]
Maybe the Serializer isn't the best fit here ? I found myself shifting some data structures from the binary serializer to the SQLite db (e.g. storing `Hash` as strings) so that it's easier to debug. The serializer was only intended for blocks and network communication, originally.

File: source/agora/flash/Channel.d (hunk 3 above)
Comment (AndrejMitrovic) [reply source/agora/flash/Channel.d:2046]: I forgot about UDAs https://github.com/bosagora/agora/pull/1815#issuecomment-803796943

**Comment by AndrejMitrovic** [comment]
> Maybe the Serializer isn't the best fit here ? I found myself shifting some data structures from the binary serializer to the SQLite db (e.g. storing Hash as strings) so that it's easier to debug. The serializer was only intended for blocks and network communication, originally.

Mmmm good point. I just stored everything as a blob and that does makes it hard to debug later. I think turning structs into their own tables is gonna take a little more effort though. How about I file an issue and we fix this up later?

**Comment by AndrejMitrovic** [comment]
The failure seems to be unrelated to this PR. The timeouts happen in other tests but only Flash code was touched.

**Comment by AndrejMitrovic** [comment]
Finally it's green.

**Comment by AndrejMitrovic** [comment]
@omerfirmak would you consider merging this? It's blocking some other work. :>

**Review by omerfirmak — APPROVED** [review]

**Pull request merged**